lib/src/solver/failure.dart

```diff
@@ -19,6 +23,294 @@ class SolveFailure implements ApplicationException {
     assert(incompatibility.terms.single.package.isRoot);
   }
 
-  // TODO(nweiz): Produce a useful error message.
-  String toString() => "Tough luck, Chuck!";
+  /// Describes how [incompatibility] was derived, and thus why version solving
+  /// failed.
+  ///
+  /// See https://github.com/dart-lang/pub/tree/master/doc/solver.md#error-reporting
+  /// for details on how this algorithm works.
+  String toString() {
```

Comment:
This is quite the method! Given that it has state and a number of sub-functions with their own doc comments, I think it's worth moving it out to a separate class that this calls.

Reply:
Done.